lib/spack/spack/build_environment.py

```diff
@@ -1014,26 +1014,33 @@ def add_modifications_for_dep(dep):
 
 
 def get_cmake_prefix_path(pkg):
-    # Note that unlike modifications_from_dependencies, this does not include
-    # any edits to CMAKE_PREFIX_PATH defined in custom
-    # setup_dependent_build_environment implementations of dependency packages
     build_deps      = set(pkg.spec.dependencies(deptype=('build', 'test')))
     link_deps       = set(pkg.spec.traverse(root=False, deptype=('link')))
     build_link_deps = build_deps | link_deps
     spack_built = []
     externals = []
+
     # modifications_from_dependencies updates CMAKE_PREFIX_PATH by first
     # prepending all externals and then all non-externals
     for dspec in pkg.spec.traverse(root=False, order='post'):
         if dspec in build_link_deps:
             if dspec.external:
-                externals.insert(0, dspec)
+                externals.insert(0, dspec.prefix)
             else:
-                spack_built.insert(0, dspec)
-
-    ordered_build_link_deps = spack_built + externals
-    build_link_prefixes = filter_system_paths(
-        x.prefix for x in ordered_build_link_deps)
+                spack_built.insert(0, dspec.prefix)
+
+    # CMAKE_PREFIX_PATH here cannot come from the outside environment under
+    # normal circumstances, it has to come from spack.  include these second
+    # after the spack_built dependencies but before externals
```

Comment:
I think this is a good PR-level commentary on the reason for these changes, but in the context of the code I think the following would be clearer:

> Collect CMAKE_PREFIX_PATH shell variable modifications from per-package implementations of `setup_build_environment`; these can then be passed explicitly to the cmake invocation.

do you agree?